### DISCUSSION TITLE
How to update plutogrid row data with Streambuilder

### DISCUSSION THREAD

**Discussion by allenlim-0627** [bosskmk/pluto_grid#187] [Q&A]
Hi, I tried to use Streambuilder to generate a list of plutoRow records, but when I try to create/update data from Firestore, I am able to get correct data inside my Streambuilder where I generate the plutoRow records, but the PlutoGrid seems like keep getting old data. Is it something to deal with the key in flutter? 

```
                 SizedBox(
                        height: constraint.maxHeight - dataPagerHeight,
                        width: constraint.maxWidth,
                        child: StreamBuilder(
                          stream: repo.getStream(),
                          builder:
                              (context, AsyncSnapshot<QuerySnapshot> snapshot) {
                            if (snapshot.hasData) {
                              List<DocumentSnapshot> items = snapshot.data.docs;
                              plutoRows = [];
                              if (items != null && items.length > 0) {
                                for (int i = 0; i < items.length; i++) {
                                  DocumentSnapshot item = items[i];
                                  PlutoRow plutoRow = PlutoRow(
                                    cells: {
                                      DataGridColumnNameConstant.productId:
                                          PlutoCell(
                                              value: item.id != null &&
                                                      item.id.isNotEmpty
                                                  ? item.id
                                                  : ""),
                                      DataGridColumnNameConstant.productName: PlutoCell(
                                          value: item.data()[
                                                          FirestoreColumnConstant
                                                              .productName] !=
                                                      null &&
                                                  item
                                                      .data()[
                                                          FirestoreColumnConstant
                                                              .productName]
                                                      .isNotEmpty
                                              ? item.data()[
                                                  FirestoreColumnConstant
                                                      .productName]
                                              : "-"),
                                      DataGridColumnNameConstant.skuCode: PlutoCell(
                                          value: item.data()[
                                                          FirestoreColumnConstant
                                                              .skuCode] !=
                                                      null &&
                                                  item
                                                      .data()[
                                                          FirestoreColumnConstant
                                                              .skuCode]
                                                      .isNotEmpty
                                              ? item.data()[
                                                  FirestoreColumnConstant
                                                      .skuCode]
                                              : "-"),
                                      DataGridColumnNameConstant.productCategory: PlutoCell(
                                          value: item.data()[
                                                          FirestoreColumnConstant
                                                              .productCategory] !=
                                                      null &&
                                                  item
                                                      .data()[
                                                          FirestoreColumnConstant
                                                              .productCategory]
                                                      .isNotEmpty
                                              ? item.data()[
                                                  FirestoreColumnConstant
                                                      .productCategory]
                                              : "-"),
                                      DataGridColumnNameConstant.initialQty: PlutoCell(
                                          value: item.data()[
                                                          FirestoreColumnConstant
                                                              .initialQty] !=
                                                      null &&
                                                  item.data()[
                                                          FirestoreColumnConstant
                                                              .initialQty] >
                                                      0
                                              ? item.data()[
                                                  FirestoreColumnConstant
                                                      .initialQty]
                                              : "-"),
                                      DataGridColumnNameConstant.currentQty: PlutoCell(
                                          value: item.data()[
                                                          FirestoreColumnConstant
                                                              .currentQty] !=
                                                      null &&
                                                  item.data()[
                                                          FirestoreColumnConstant
                                                              .currentQty] >
                                                      0
                                              ? item.data()[
                                                  FirestoreColumnConstant
                                                      .currentQty]
                                              : "-"),
                                      DataGridColumnNameConstant.lowStockAlert: PlutoCell(
                                          value: item.data()[
                                                          FirestoreColumnConstant
                                                              .lowStockAlert] !=
                                                      null &&
                                                  item.data()[
                                                          FirestoreColumnConstant
                                                              .lowStockAlert] >
                                                      0
                                              ? item.data()[
                                                  FirestoreColumnConstant
                                                      .lowStockAlert]
                                              : "-"),
                                      DataGridColumnNameConstant.stockStatus: PlutoCell(
                                          value: item.data()[
                                                          FirestoreColumnConstant
                                                              .stockStatus] !=
                                                      null &&
                                                  item
                                                      .data()[
                                                          FirestoreColumnConstant
                                                              .stockStatus]
                                                      .isNotEmpty
                                              ? item.data()[
                                                  FirestoreColumnConstant
                                                      .stockStatus]
                                              : "-"),
                                    },
                                  );
                                  plutoRows.add(plutoRow);
                                }
                              }
                              return _buildPlutoBody(context, plutoRows);
                            } else {
                              return Center(child: Text("No data."));
                            }
                          },
                        ),
                      )
```

This is the part where I insert plutoRow into plutoGrid
```
Widget _buildPlutoBody(BuildContext context, List<PlutoRow> plutoRows) {
    return Consumer(
        builder: (context, ProductSummaryRepository productRepo, _) {
      return PlutoGrid(
        columns: plutoColumns,
        rows: plutoRows,
        onLoaded: (PlutoGridOnLoadedEvent event) {
          event.stateManager.setSelectingMode(PlutoGridSelectingMode.row);
          stateManager = event.stateManager;
        },
        onSelected: (PlutoGridOnSelectedEvent event) {
          if (event.row != null) {
            _navigateToCreateEditPage(context, event.row);
          }
        },
        mode: PlutoGridMode.select,
      );
    });
  }
```

This is my repository to handle Firestore query
```
Stream<QuerySnapshot> getStream() {
    return collection
        .where(FirestoreColumnConstant.status, isEqualTo: RecordStatus.active)
        .orderBy(FirestoreColumnConstant.createdOn, descending: true)
        .snapshots();
  }
```

Thank you so much and please kindly advise on how to get the data right.

**Comment by bosskmk** [comment]
Probably because of state management.
Any further answers are difficult as I can't debug by logging in the middle of your code.

**Comment by AbdoPrDZ** [comment]
```dart
import 'package:firebase_database/firebase_database.dart';
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:pluto_grid/pluto_grid.dart';

abstract class Model {
  Map<String, dynamic> get map => {};
  Map<String, dynamic> get dataGridMap => map;

  getDataGridValue(String key) => dataGridMap[key];
}

// Model of data
class ClientModel extends Model {
  late final int id;
  late final String fullname, phone;
  late final DateTime createdAt;

  ClientModel(
    this.id,
    this.fullname,
    this.phone,
    this.createdAt,
  );

  static DatabaseReference ref = FirebaseDatabase.instance.ref('clients');

  static Future<ClientModel?> fromId(int id) async {
    Map? data = (await ref.child('$id').get()).value as Map?;
    return data != null ? fromMap(data) : null;
  }

  static Future<List<ClientModel>> all() async {
    List? data = (await ref.get()).value as List?;
    return data != null ? allFromMap(List<Map>.from(data)) : [];
  }

  static Stream<List<ClientModel>> stream() =>
      ref.onValue.asyncExpand<List<ClientModel>>(
        (event) async* {
          if (event.snapshot.value != null) {
            List data = event.snapshot.value as List;
            if (data.isNotEmpty) {
              yield allFromMap(data);
            }
          }
        },
      );

  static List<ClientModel> allFromMap(List items) =>
      [for (Map data in items) fromMap(data)];

  static ClientModel fromMap(Map data) => ClientModel(
        data['id'],
        data['fullname'],
        data['phone'],
        DateTime.fromMillisecondsSinceEpoch(data['created_at']),
      );

  Map<String, dynamic> get map => {
        'id': id,
        'fullname': fullname,
        'phone': phone,
        'created_at': createdAt.millisecondsSinceEpoch,
      };

  static Future<int> getNewId() async =>
      ((await ref.get()).value as List?)?.length ?? 0;

  static Future<ClientModel> create(Map data) async {
    ClientModel item = ClientModel.fromMap(data);
    await ref.child('${item.id}').set(item.map);
    return item;
  }

  update(Map data) async {
    fullname = data['fullname'] ?? fullname;
    phone = data['phone'] ?? phone;
    createdAt = data['created_at'] ?? createdAt;
    await ref.child('$id').update(map);
  }

  delete() async => remove(id);

  static Future remove(int id) => ref.child('$id').remove();
}

// Table Widget
class TableView extends StatelessWidget {
  final TableViewController controller;
  final bool Function(Model model)? avaliableModels;
  final EdgeInsets? margin, padding;

  const TableView({
    required this.controller,
    this.avaliableModels,
    this.margin,
    this.padding,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return PlutoGrid(
      columns: controller.stateManager.columns,
      // rows: rows,
      rows: controller.stateManager.rows,
      onChanged: (PlutoGridOnChangedEvent event) {
        print(event);
      },
      onLoaded: (PlutoGridOnLoadedEvent event) {
        controller.stateManager = event.stateManager;
        controller.stateManager.setSelectingMode(PlutoGridSelectingMode.cell);
        controller.stateManager.setShowColumnFilter(true);
      },
      configuration: const PlutoGridConfiguration(),
    );
  }
}

class TableViewController {
  final Stream<List<Model>>? streamSource;
  final LoadAction streamLoadAction;
  final List<PlutoColumn> columns;
  final bool Function(Model model)? avaliableModels;

  late PlutoGridStateManager stateManager;

  TableViewController({
    List<Model> initialItems = const [],
    this.streamSource,
    this.streamLoadAction = LoadAction.replace,
    required this.columns,
    this.avaliableModels,
  }) {
    items = initialItems;
    if (streamSource != null) {
      streamSource!.listen((items) => loadItems(
            items,
            loadAction: streamLoadAction,
          ));
      stateManager = PlutoGridStateManager(
        columns: columns,
        rows: [],
        gridFocusNode: null,
        scroll: null,
      );
      update();
    }
  }

  List<String> get fields => [for (PlutoColumn column in columns) column.field];

  List<Model> items = [];

  List<Model> filterModels(List<Model> items) {
    List<Model> filterdModels = [];
    for (Model item in items) {
      if (avaliableModels!(item)) filterdModels.add(item);
    }
    return filterdModels;
  }

  loadItems(List<Model> items, {LoadAction loadAction = LoadAction.replace}) {
    items = avaliableModels != null ? filterModels(items) : items;
    switch (loadAction) {
      case LoadAction.addAfter:
        this.items.addAll(items);
        break;
      case LoadAction.addBefor:
        items.addAll(this.items);
        this.items = items;
        break;
      default:
        this.items = items;
    }
    update();
  }

  update() async {
    stateManager.setShowLoading(true);
    List<PlutoRow> refRows = [
      for (Model model in items)
        PlutoRow(
          cells: {
            for (String field in fields)
              field: PlutoCell(value: model.getDataGridValue(field))
          },
        )
    ];
    stateManager.refRows.clear();
    stateManager.refRows.addAll(refRows);
    stateManager.setShowLoading(false);
  }
}

class LoadAction {
  final String value;
  const LoadAction(this.value);

  static const LoadAction replace = LoadAction('replace');
  static const LoadAction addAfter = LoadAction('add-after');
  static const LoadAction addBefor = LoadAction('add-befor');
}

// usage

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _HomePageState();
}

class _HomePageState extends State<MyHomePage> {
  late TextEditingController fullnameController;
  late TextEditingController phoneController;

  late TableViewController tableViewController;

  @override
  void initState() {
    fullnameController = TextEditingController();
    phoneController = TextEditingController();

    tableViewController = TableViewController(
      streamSource: ClientModel.stream(), // stream from firebase database
      avaliableModels: (model) {
        ClientModel client = model as ClientModel;
        //  filter models code
        return client.createdAt.year >= 2022;
      },
      columns: [
        // columns you want to show
        PlutoColumn(
          title: '#',
          field: 'id', // must be equals with datagrid map keys
          type: PlutoColumnType.number(),
        ),
        PlutoColumn(
          title: 'Fullname',
          field: 'fullname',
          type: PlutoColumnType.text(),
        ),
        PlutoColumn(
          title: 'Phone',
          field: 'phone',
          type: PlutoColumnType.text(),
        ),
      ],
    );

    super.initState();
  }

  @override
  void dispose() {
    fullnameController.dispose();
    phoneController.dispose();
    super.dispose();
  }

  addClient() async {
    if (fullnameController.text.isEmpty || phoneController.text.isEmpty) {
      print('please fill all data');
      return;
    }
    // 'isPhoneNumber' from Getx package
    else if (!phoneController.text.isPhoneNumber) {
      print('Invalid phone number');
      return;
    }
    await ClientModel.create({
      'id': await ClientModel.getNewId(),
      'fullname': fullnameController.text,
      'phone': phoneController.text,
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SizedBox(
        width: MediaQuery.of(context).size.width,
        height: MediaQuery.of(context).size.height,
        child: Flex(
          direction: Axis.vertical,
          children: [
            TextField(
              controller: fullnameController,
              decoration: const InputDecoration(hintText: 'Fullname'),
            ),
            TextField(
              controller: phoneController,
              decoration: const InputDecoration(hintText: 'Phone'),
            ),
            ElevatedButton(onPressed: addClient, child: const Text('add')),
            Flexible(child: TableView(controller: tableViewController))
          ],
        ),
      ),
    );
  }
}
```

**Reply by cbotelho** [reply]
> ```dart
> import 'package:firebase_database/firebase_database.dart';
> import 'package:flutter/material.dart';
> import 'package:get/get.dart';
> import 'package:pluto_grid/pluto_grid.dart';
> 
> abstract class Model {
>   Map<String, dynamic> get map => {};
>   Map<String, dynamic> get dataGridMap => map;
> 
>   getDataGridValue(String key) => dataGridMap[key];
> }
> 
> // Model of data
> class ClientModel extends Model {
>   late final int id;
>   late final String fullname, phone;
>   late final DateTime createdAt;
> 
>   ClientModel(
>     this.id,
>     this.fullname,
>     this.phone,
>     this.createdAt,
>   );
> 
>   static DatabaseReference ref = FirebaseDatabase.instance.ref('clients');
> 
>   static Future<ClientModel?> fromId(int id) async {
>     Map? data = (await ref.child('$id').get()).value as Map?;
>     return data != null ? fromMap(data) : null;
>   }
> 
>   static Future<List<ClientModel>> all() async {
>     List? data = (await ref.get()).value as List?;
>     return data != null ? allFromMap(List<Map>.from(data)) : [];
>   }
> 
>   static Stream<List<ClientModel>> stream() =>
>       ref.onValue.asyncExpand<List<ClientModel>>(
>         (event) async* {
>           if (event.snapshot.value != null) {
>             List data = event.snapshot.value as List;
>             if (data.isNotEmpty) {
>               yield allFromMap(data);
>             }
>           }
>         },
>       );
> 
>   static List<ClientModel> allFromMap(List items) =>
>       [for (Map data in items) fromMap(data)];
> 
>   static ClientModel fromMap(Map data) => ClientModel(
>         data['id'],
>         data['fullname'],
>         data['phone'],
>         DateTime.fromMillisecondsSinceEpoch(data['created_at']),
>       );
> 
>   Map<String, dynamic> get map => {
>         'id': id,
>         'fullname': fullname,
>         'phone': phone,
>         'created_at': createdAt.millisecondsSinceEpoch,
>       };
> 
>   static Future<int> getNewId() async =>
>       ((await ref.get()).value as List?)?.length ?? 0;
> 
>   static Future<ClientModel> create(Map data) async {
>     ClientModel item = ClientModel.fromMap(data);
>     await ref.child('${item.id}').set(item.map);
>     return item;
>   }
> 
>   update(Map data) async {
>     fullname = data['fullname'] ?? fullname;
>     phone = data['phone'] ?? phone;
>     createdAt = data['created_at'] ?? createdAt;
>     await ref.child('$id').update(map);
>   }
> 
>   delete() async => remove(id);
> 
>   static Future remove(int id) => ref.child('$id').remove();
> }
> 
> // Table Widget
> class TableView extends StatelessWidget {
>   final TableViewController controller;
>   final bool Function(Model model)? avaliableModels;
>   final EdgeInsets? margin, padding;
> 
>   const TableView({
>     required this.controller,
>     this.avaliableModels,
>     this.margin,
>     this.padding,
>     Key? key,
>   }) : super(key: key);
> 
>   @override
>   Widget build(BuildContext context) {
>     return PlutoGrid(
>       columns: controller.stateManager.columns,
>       // rows: rows,
>       rows: controller.stateManager.rows,
>       onChanged: (PlutoGridOnChangedEvent event) {
>         print(event);
>       },
>       onLoaded: (PlutoGridOnLoadedEvent event) {
>         controller.stateManager = event.stateManager;
>         controller.stateManager.setSelectingMode(PlutoGridSelectingMode.cell);
>         controller.stateManager.setShowColumnFilter(true);
>       },
>       configuration: const PlutoGridConfiguration(),
>     );
>   }
> }
> 
> class TableViewController {
>   final Stream<List<Model>>? streamSource;
>   final LoadAction streamLoadAction;
>   final List<PlutoColumn> columns;
>   final bool Function(Model model)? avaliableModels;
> 
>   late PlutoGridStateManager stateManager;
> 
>   TableViewController({
>     List<Model> initialItems = const [],
>     this.streamSource,
>     this.streamLoadAction = LoadAction.replace,
>     required this.columns,
>     this.avaliableModels,
>   }) {
>     items = initialItems;
>     if (streamSource != null) {
>       streamSource!.listen((items) => loadItems(
>             items,
>             loadAction: streamLoadAction,
>           ));
>       stateManager = PlutoGridStateManager(
>         columns: columns,
>         rows: [],
>         gridFocusNode: null,
>         scroll: null,
>       );
>       update();
>     }
>   }
> 
>   List<String> get fields => [for (PlutoColumn column in columns) column.field];
> 
>   List<Model> items = [];
> 
>   List<Model> filterModels(List<Model> items) {
>     List<Model> filterdModels = [];
>     for (Model item in items) {
>       if (avaliableModels!(item)) filterdModels.add(item);
>     }
>     return filterdModels;
>   }
> 
>   loadItems(List<Model> items, {LoadAction loadAction = LoadAction.replace}) {
>     items = avaliableModels != null ? filterModels(items) : items;
>     switch (loadAction) {
>       case LoadAction.addAfter:
>         this.items.addAll(items);
>         break;
>       case LoadAction.addBefor:
>         items.addAll(this.items);
>         this.items = items;
>         break;
>       default:
>         this.items = items;
>     }
>     update();
>   }
> 
>   update() async {
>     stateManager.setShowLoading(true);
>     List<PlutoRow> refRows = [
>       for (Model model in items)
>         PlutoRow(
>           cells: {
>             for (String field in fields)
>               field: PlutoCell(value: model.getDataGridValue(field))
>           },
>         )
>     ];
>     stateManager.refRows.clear();
>     stateManager.refRows.addAll(refRows);
>     stateManager.setShowLoading(false);
>   }
> }
> 
> class LoadAction {
>   final String value;
>   const LoadAction(this.value);
> 
>   static const LoadAction replace = LoadAction('replace');
>   static const LoadAction addAfter = LoadAction('add-after');
>   static const LoadAction addBefor = LoadAction('add-befor');
> }
> 
> // usage
> 
> class MyHomePage extends StatefulWidget {
>   const MyHomePage({super.key});
> 
>   @override
>   State<MyHomePage> createState() => _HomePageState();
> }
> 
> class _HomePageState extends State<MyHomePage> {
>   late TextEditingController fullnameController;
>   late TextEditingController phoneController;
> 
>   late TableViewController tableViewController;
> 
>   @override
>   void initState() {
>     fullnameController = TextEditingController();
>     phoneController = TextEditingController();
> 
>     tableViewController = TableViewController(
>       streamSource: ClientModel.stream(), // stream from firebase database
>       avaliableModels: (model) {
>         ClientModel client = model as ClientModel;
>         //  filter models code
>         return client.createdAt.year >= 2022;
>       },
>       columns: [
>         // columns you want to show
>         PlutoColumn(
>           title: '#',
>           field: 'id', // must be equals with datagrid map keys
>           type: PlutoColumnType.number(),
>         ),
>         PlutoColumn(
>           title: 'Fullname',
>           field: 'fullname',
>           type: PlutoColumnType.text(),
>         ),
>         PlutoColumn(
>           title: 'Phone',
>           field: 'phone',
>           type: PlutoColumnType.text(),
>         ),
>       ],
>     );
> 
>     super.initState();
>   }
> 
>   @override
>   void dispose() {
>     fullnameController.dispose();
>     phoneController.dispose();
>     super.dispose();
>   }
> 
>   addClient() async {
>     if (fullnameController.text.isEmpty || phoneController.text.isEmpty) {
>       print('please fill all data');
>       return;
>     }
>     // 'isPhoneNumber' from Getx package
>     else if (!phoneController.text.isPhoneNumber) {
>       print('Invalid phone number');
>       return;
>     }
>     await ClientModel.create({
>       'id': await ClientModel.getNewId(),
>       'fullname': fullnameController.text,
>       'phone': phoneController.text,
>     });
>   }
> 
>   @override
>   Widget build(BuildContext context) {
>     return Scaffold(
>       body: SizedBox(
>         width: MediaQuery.of(context).size.width,
>         height: MediaQuery.of(context).size.height,
>         child: Flex(
>           direction: Axis.vertical,
>           children: [
>             TextField(
>               controller: fullnameController,
>               decoration: const InputDecoration(hintText: 'Fullname'),
>             ),
>             TextField(
>               controller: phoneController,
>               decoration: const InputDecoration(hintText: 'Phone'),
>             ),
>             ElevatedButton(onPressed: addClient, child: const Text('add')),
>             Flexible(child: TableView(controller: tableViewController))
>           ],
>         ),
>       ),
>     );
>   }
> }
> ```

Excelente código..
E como eu poderia pegar esses dados de um controller/model:
Entidade_controller.dart:
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:infodoc_cloud/app/core/repositories/entidade_repository.dart';
import 'package:infodoc_cloud/app/models/entidade_model.dart';
import 'package:infodoc_cloud/app/theme/app_colors.dart';

enum Acao {
  list,
  create,
  update,
  detail,
}
class EntidadeController extends GetxController {
  final EntidadeRepository repository;
  EntidadeController({required this.repository});
  @override
  void onInit() async {
    await getAll();
    super.onInit();
  }
  var id = "";
  var nome = "";
  var cnpj = "";
  var descricao = "";
  var telefone = "";
  var titulo = "Listar Entidades".obs;
  var tela = Acao.list.obs;
  var entidades = <EntidadeModel>[].obs;
  EntidadeModel? entidadeSelected;
  void changeAction(value) {
    tela.value = value;
    if (value == Acao.list) {
      titulo.value = "Lista de Entidades";
    } else if (value == Acao.create) {
      titulo.value = "Nova Entidade";
      id = "";
      nome = "";
      descricao = "";
      cnpj = "";
      telefone = "";
    } else if (value == Acao.detail) {
      titulo.value = "Detalhe de Entidade";
    } else if (value == Acao.update) {
      titulo.value = "Alterar a Entidade";
      nome = entidadeSelected!.nome!;
      descricao = entidadeSelected!.descricao!;
      cnpj = entidadeSelected!.cnpj!;
      telefone = entidadeSelected!.telefone!;
    } else {
      titulo.value = "Alterar de Entidade";
    }
  }
  Future<void> getAll() async {
    entidades.value = await repository.getAll();
  }
  Future<void> criar() async {
    await repository.criar(
      id: id,
      nome: nome,
      descricao: descricao, 
      cnpj: cnpj,
      telefone: telefone,

      ).then((value) {
      getAll();
      Get.defaultDialog(
        title: ('Sucesso!'),
        titleStyle: const TextStyle(color: AppColors.textColor, fontFamily: 'ROBOTO', fontSize: 15, fontWeight: FontWeight.bold),
        radius: 8,
        middleText: 'Entidade Cadastrada Com Sucesso!',
        backgroundColor: const Color(0xFFF1F9FC),
        actions: [TextButton(
          style: TextButton.styleFrom(
            backgroundColor: const Color(0xFF028506), 
            foregroundColor: Colors.white, 
            padding: const EdgeInsets.all(14.0), 
            textStyle: const TextStyle(fontFamily: 'ROBOTO', fontSize: 12)),
          onPressed: () => Get.back(), 
          child: const Text('Fechar')
          
          )]
      );
      tela.value = Acao.list;
    });
  }
  Future<void> remover(id) async {
    await repository.remover(id).then((value) {
      getAll();
      Get.defaultDialog(
        title: ('Alerta!'),
        titleStyle: const TextStyle(color: AppColors.textColor, fontFamily: 'ROBOTO', fontSize: 15, fontWeight: FontWeight.bold),
        radius: 8,
        middleText: 'Entidade Excluida Com Sucesso!',
        backgroundColor: const Color(0xFFF1F9FC),
        actions: [TextButton(
          style: TextButton.styleFrom(
            backgroundColor: const Color(0xFFF10B03), 
            foregroundColor: Colors.white, 
            padding: const EdgeInsets.all(14.0), 
            textStyle: const TextStyle(fontFamily: 'ROBOTO', fontSize: 12)),
          onPressed: () => Get.back(), 
          child: const Text('Fechar')  
          )]
      );
      tela.value = Acao.list;
    });
  }
  Future<void> alterar() async {
    entidadeSelected?.nome = nome;
    entidadeSelected?.descricao = descricao;
    entidadeSelected?.cnpj = cnpj;
    entidadeSelected?.telefone  = telefone;
    await repository.alterar(entidadeSelected!).then((value) {
      getAll();
      Get.defaultDialog(
        title: ('Sucesso!'),
        titleStyle: const TextStyle(color: AppColors.textColor, fontFamily: 'ROBOTO', fontSize: 15, fontWeight: FontWeight.bold),
        radius: 8,
        middleText: 'Entidade Alterado Com Sucesso!',
        backgroundColor: const Color(0xFFF1F9FC),
        actions: [TextButton(
          style: TextButton.styleFrom(
            backgroundColor: const Color(0xFF026ECD), 
            foregroundColor: Colors.white, 
            padding: const EdgeInsets.all(14.0), 
            textStyle: const TextStyle(fontFamily: 'ROBOTO', fontSize: 12)),
          onPressed: () => Get.back(), 
          child: const Text('Fechar' )
]
      );
     tela.value = Acao.list;
    });
  }
  void changeNome(value) {nome = value;}
  void changeDescricao(value) {nome = value; }
  void changeCnpj(value){cnpj = value; } 
void changeTelefone(value){telefone = value;} 
void showDetails(index) {
    entidadeSelected = entidades.value[index];
    tela.value = Acao.detail;
  }

and the model page:
entity_model.dart:
class ModelEntity {
   String? id;
   String? Name;
   String? description;
   String? CNPJ;
   String? telephone;
  

   ModelEntity({
     required this.id,
     required this.name,
     required this.description,
     required this.cnpj,
     this.phone,
  
     });

   EntidadeModel.fromJson(Map<String, dynamic> json, String uid) {
     json.addAll({'uid': uid});
     id = json['uid'];
     name = json['name'];
     description = json['description'];
     cnpj = json['cnpj'];
     phone = json['phone'];
 
   }

   Map<String, dynamic> toJson() {
     final Map<String, dynamic> data = Map<String, dynamic>();
     data['uid'] = id;
     data['name'] = name;
     data['description'] = description;
     data['cnpj'] = cnpj;
     data['phone'] = phone;
   return data;
   }
}
thanks for the help.

**Comment by michaelbushe** [comment]
My bug might help you:
https://github.com/bosskmk/pluto_grid/issues/828

The bug points to my working example with firebase and Pluto Grid.  
https://github.com/michaelbushe/pluto_grid_riverpod_update_example

Note that the workaround for this bug was to have the PlutoGrid(key: ObjectKey(rows)...)
